core-bundle/src/Routing/Candidates/AbstractCandidates.php

```diff
@@ -0,0 +1,161 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Routing\Candidates;
+
+use Symfony\Cmf\Component\Routing\Candidates\CandidatesInterface;
+use Symfony\Component\HttpFoundation\Request;
+
+class AbstractCandidates implements CandidatesInterface
+{
+    /**
+     * A limit to apply to the number of candidates generated.
+     *
+     * This is to prevent abusive requests with a lot of "/". The limit is per
+     * batch, that is if a locale matches you could get as many as 2 * $limit
+     * candidates if the URL has that many slashes.
+     *
+     * @var int
+     */
+    private const LIMIT = 20;
+
+    /**
+     * @var array
+     */
+    private $urlPrefixes;
+
+    /**
+     * @var array
+     */
+    private $urlSuffixes;
+
+    public function __construct(array $urlPrefixes, array $urlSuffixes)
+    {
+        $this->urlPrefixes = $urlPrefixes;
+        $this->urlSuffixes = $urlSuffixes;
+    }
+
+    public function isCandidate($name): bool
+    {
+        return 0 === strncmp($name, 'tl_page.', 8);
+    }
+
+    public function restrictQuery($queryBuilder): void
+    {
+    }
+
+    /**
+     * Generates possible page aliases from the request path by
+     * removing prefixes, suffixes and parameters.
+     *
+     * Example 1:
+     *   Path: /en/alias/foo/bar.html
+     *   Prefixes: [en, de]
+     *   Suffixes: [.html]
+     *   Possible aliases:
+     *     - alias/foo/bar
+     *     - alias/foo
+     *     - alias
+     *
+     * Example 2:
+     *   Path: /en/alias/foo/bar.html
+     *   Prefixes: [en, '']
+     *   Suffixes: [.html, '']
+     *   Possible aliases:
+     *     - en/alias/foo/bar.html
+     *     - en/alias/foo/bar
+     *     - en/alias/foo
+     *     - en/alias
+     *     - alias/foo/bar.html
+     *     - alias/foo/bar
+     *     - alias/foo
+     *     - alias
+     */
```

Comment:
If I try example 2, I'll get the following (also including `en`):

    Array
    (
        [0] => alias/foo/bar
        [1] => alias/foo
        [2] => alias
        [3] => alias/foo/bar.html
        [4] => en/alias/foo/bar
        [5] => en/alias/foo
        [6] => en/alias
        [7] => en
        [8] => en/alias/foo/bar.html
    )